src/main/kotlin/ee/carlrobert/codegpt/codecompletions/CodeGPTInlineCompletionProvider.kt

```diff
@@ -75,9 +78,19 @@ class CodeGPTInlineCompletionProvider : InlineCompletionProvider {
 
     class CodeCompletionEventListener(
         private val requestDetails: InfillRequestDetails,
-        private val completed: (String) -> Unit
+        private val completed: (String, Boolean) -> Unit
     ) : CompletionEventListener<String> {
 
+        private var isStreaming = false
+
+        override fun onMessage(message: String?, eventSource: EventSource?) {
+            if (message != null) {
+                isStreaming = true
+                val needCancel = message.contains('\n')
+                completed(message.takeWhile { it != '\n' }, needCancel)
```

Comment:
This doesn't actually cancel the request (and save tokens), rather it just cancels early from the callback.

Would something like this work:

```kt
private val messageBuilder = StringBuilder()

override fun onMessage(message: String?, eventSource: EventSource?) {
    messageBuilder.append(message)
    if (messageBuilder.contains('\n')) {
        eventSource?.cancel()
    }
}
```

now, once `cancel()` is called, the `onComplete` callback should be triggered without requiring any other changes

Reply:
you could also trim the result down to the first element before the newline as well

Reply:
But this has been verified through my testing and the Stream was indeed cancelled early. The llama server local immediately stopped the inferencing.
Below are the logs I printed in the onMessage() and cancelCurrentCall() functions.

```
[OkHttp http://localhost:8080/...] INFO ee.carlrobert.llm.completion.CompletionEventSourceListener - Request opened.
2024-04-16 08:54:39,755 [  75441]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - onMessage:public
2024-04-16 08:54:39,795 [  75481]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - onMessage: static
2024-04-16 08:54:39,834 [  75520]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - onMessage: void
2024-04-16 08:54:39,877 [  75563]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - onMessage: main
2024-04-16 08:54:39,919 [  75605]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - onMessage:(
2024-04-16 08:54:39,960 [  75646]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - onMessage:String
2024-04-16 08:54:40,001 [  75687]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - onMessage:[]
2024-04-16 08:54:40,041 [  75727]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - onMessage: args
2024-04-16 08:54:40,082 [  75768]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - onMessage:)
2024-04-16 08:54:40,123 [  75809]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - onMessage: {
2024-04-16 08:54:40,163 [  75849]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - onMessage:

2024-04-16 08:54:40,164 [  75850]   WARN - #ee.carlrobert.codegpt.codecompletions.CodeGPTInlineCompletionProvider - cancelCurrentCall called!
[OkHttp http://localhost:8080/...] INFO ee.carlrobert.llm.completion.CompletionEventSourceListener - Stream was cancelled
```

Reply:
Ah right, `cancelCurrentCall` is invoked based on the `cancel` condition. Nevertheless, I would still like to see the cancelling being done inside the `onMessage` callback, then you don't need to declare any redundant class variables etc.

Reply:
I think you are right, updated and verified. 
and I think no action needed in onCancelled(), Code is more concise now.